### ISSUE TITLE
Jupyter terminal jobs never quit

### ISSUE THREAD

**Issue by tbrodbeck** [jupyterlab/jupyterlab#10447]
## Description

I realized that jobs in the jupyter terminal jobs do quit when the terminal is closed (what can happend automatically for some reason).  This leads to a lot of process bloat that gets created over time eating up memory and cpu.

## Reproduce

1. Open a jupyter server.
2. Create a Terminal.
3. Execute a command. For example `htop` (the same applies to `bash` or `zsh` itself btw)
4. Close the Terminal
5. Check processes hosting machine. `htop` does not quit

## Expected behavior
Jobs should quit when the Terminal hosting them is gone (and the job was not disowned). This behaviour would be consistent with normal local terminals and would avoid keeping jobs running without knowledge about that.

## Context

- Operating System and version: OS: Ubuntu 20.04.2 LTS x86_64
- Browser and version: Brave Version 1.25.72 Chromium: 91.0.4472.101 (Official Build) (x86_64) (for example)
- JupyterLab version: 3.0.16

<details><summary>Troubleshoot Output</summary>
<pre>
$PATH:
	/usr/local/cuda-11.3/bin
	/home/linuxbrew/.linuxbrew/bin
	/home/linuxbrew/.linuxbrew/sbin
	/usr/local/sbin
	/usr/local/bin
	/usr/sbin
	/usr/bin
	/sbin
	/bin
	/usr/games
	/usr/local/games
	/snap/bin
	/home/t/res/bin
	/home/t/.local/bin
	/home/linuxbrew/.linuxbrew/opt/fzf/bin

sys.path:
	/home/t/.local/bin
	/usr/lib/python38.zip
	/usr/lib/python3.8
	/usr/lib/python3.8/lib-dynload
	/home/t/.local/lib/python3.8/site-packages
	/usr/local/lib/python3.8/dist-packages
	/usr/lib/python3/dist-packages

sys.executable:
	/usr/bin/python3

sys.version:
	3.8.5 (default, May 27 2021, 13:30:53)
	[GCC 9.3.0]

platform.platform():
	Linux-5.4.0-74-generic-x86_64-with-glibc2.29

which -a jupyter:
	/home/t/.local/bin/jupyter

pip list:
	Package                   Version
	------------------------- --------------------
	absl-py                   0.12.0
	acme                      1.1.0
	alembic                   1.4.1
	anyio                     3.1.0
	argon2-cffi               20.1.0
	astunparse                1.6.3
	async-generator           1.10
	attrdict                  2.0.1
	attrs                     21.2.0
	Automat                   0.8.0
	Babel                     2.9.1
	backcall                  0.2.0
	bleach                    3.3.0
	blinker                   1.4
	blis                      0.4.1
	bravado                   11.0.3
	bravado-core              5.17.0
	cachetools                4.2.2
	catalogue                 1.0.0
	certbot                   0.40.0
	certbot-nginx             0.40.0
	certifi                   2019.11.28
	cffi                      1.14.5
	chardet                   3.0.4
	Click                     7.0
	cloud-init                21.1
	cloud-tpu-client          0.10
	cloudpickle               1.6.0
	colorama                  0.4.3
	command-not-found         0.3
	ConfigArgParse            0.13.0
	configobj                 5.0.6
	constantly                15.1.0
	cryptography              2.8
	cupshelpers               1.0
	cycler                    0.10.0
	cymem                     2.0.3
	databricks-cli            0.14.3
	dbus-python               1.2.16
	decorator                 5.0.9
	defer                     1.0.6
	defusedxml                0.7.1
	distro                    1.4.0
	distro-info               0.23ubuntu1
	dnspython                 1.16.0
	docker                    5.0.0
	docopt                    0.6.2
	entrypoints               0.3
	et-xmlfile                1.0.1
	fail2ban                  0.11.1
	filelock                  3.0.12
	Flask                     2.0.0
	flatbuffers               1.12
	future                    0.18.2
	gast                      0.4.0
	gitdb                     4.0.7
	GitPython                 3.1.17
	google-api-core           1.30.0
	google-api-python-client  1.8.0
	google-auth               1.30.0
	google-auth-httplib2      0.1.0
	google-auth-oauthlib      0.4.4
	google-pasta              0.2.0
	googleapis-common-protos  1.53.0
	gpg                       1.13.1-unknown
	greenlet                  0.4.15
	grpcio                    1.34.1
	gunicorn                  20.1.0
	h5py                      3.1.0
	httplib2                  0.14.0
	hyperlink                 19.0.0
	idna                      2.8
	importlib-metadata        1.5.0
	incremental               16.10.1
	ipykernel                 5.5.5
	ipython                   7.24.1
	ipython-genutils          0.2.0
	ipywidgets                7.6.3
	itsdangerous              2.0.1
	jdcal                     1.4.1
	jedi                      0.18.0
	Jinja2                    3.0.1
	joblib                    0.16.0
	josepy                    1.2.0
	json5                     0.9.5
	jsonpatch                 1.22
	jsonpickle                1.5.2
	jsonpointer               2.0
	jsonref                   0.2
	jsonschema                3.2.0
	jupyter                   1.0.0
	jupyter-client            6.1.12
	jupyter-console           6.4.0
	jupyter-core              4.7.1
	jupyter-server            1.8.0
	jupyterlab                3.0.16
	jupyterlab-pygments       0.1.2
	jupyterlab-server         2.5.2
	jupyterlab-widgets        1.0.0
	kaggle                    1.5.12
	Keras                     2.4.3
	keras-nightly             2.5.0.dev2021032900
	Keras-Preprocessing       1.1.2
	keyring                   18.0.1
	kiwisolver                1.2.0
	language-selector         0.1
	launchpadlib              1.10.13
	lazr.restfulclient        0.14.2
	lazr.uri                  1.0.3
	macaroonbakery            1.3.1
	Mako                      1.1.4
	Markdown                  3.1.1
	MarkupSafe                2.0.1
	matplotlib                3.2.2
	matplotlib-inline         0.1.2
	mistune                   0.8.4
	mlflow                    1.17.0
	mock                      3.0.5
	monotonic                 1.6
	more-itertools            4.2.0
	mpmath                    1.1.0
	msgpack                   0.6.2
	munch                     2.5.0
	murmurhash                1.0.2
	nbclassic                 0.3.1
	nbclient                  0.5.3
	nbconvert                 6.0.7
	nbformat                  5.1.3
	neptune-client            0.9.16
	neptune-contrib           0.27.1
	nest-asyncio              1.5.1
	netifaces                 0.10.4
	nltk                      3.5
	notebook                  6.4.0
	numpy                     1.19.5
	oauth2client              4.1.3
	oauthlib                  3.1.0
	olefile                   0.46
	openpyxl                  3.0.5
	opt-einsum                3.3.0
	packaging                 20.9
	pandas                    1.1.2
	pandocfilters             1.4.3
	parsedatetime             2.4
	parso                     0.8.2
	pbr                       5.4.5
	pexpect                   4.8.0
	pickleshare               0.7.5
	Pillow                    7.0.0
	pip                       21.1.2
	plac                      1.1.3
	plotly                    4.14.3
	preshed                   3.0.2
	prometheus-client         0.11.0
	prometheus-flask-exporter 0.18.2
	prompt-toolkit            3.0.19
	protobuf                  3.17.0
	psutil                    5.5.1
	ptyprocess                0.7.0
	py-cpuinfo                8.0.0
	pyasn1                    0.4.2
	pyasn1-modules            0.2.1
	pycairo                   1.16.2
	pycparser                 2.20
	pycrypto                  2.6.1
	pycups                    1.9.73
	Pygments                  2.9.0
	PyGObject                 3.36.0
	PyHamcrest                1.9.0
	PyICU                     2.4.2
	pyinotify                 0.9.6
	PyJWT                     1.7.1
	pymacaroons               0.13.0
	PyNaCl                    1.3.0
	pynvim                    0.4.1
	pyOpenSSL                 19.0.0
	pyparsing                 2.4.7
	pyRFC3339                 1.1
	pyrsistent                0.17.3
	pyserial                  3.4
	python-apt                2.0.0+ubuntu0.20.4.4
	python-dateutil           2.8.1
	python-debian             0.1.36ubuntu1
	python-decouple           3.4
	python-editor             1.0.4
	python-slugify            5.0.2
	pytz                      2020.1
	PyYAML                    5.3.1
	pyzmq                     22.1.0
	qtconsole                 5.1.0
	QtPy                      1.9.0
	querystring-parser        1.2.4
	ranger-fm                 1.9.3
	regex                     2020.7.14
	requests                  2.22.0
	requests-oauthlib         1.3.0
	requests-toolbelt         0.8.0
	requests-unixsocket       0.2.0
	retrying                  1.3.3
	rfc3987                   1.3.8
	rouge                     1.0.0
	rsa                       4.7.2
	s-tui                     1.0.0b3
	sacred                    0.8.2
	sacremoses                0.0.43
	scikit-learn              0.23.1
	scipy                     1.5.2
	screen-resolution-extra   0.0.0
	SecretStorage             2.3.1
	Send2Trash                1.5.0
	sentence-transformers     0.3.6
	sentencepiece             0.1.91
	service-identity          18.1.0
	setuptools                57.0.0
	simplejson                3.16.0
	six                       1.16.0
	smmap                     4.0.0
	sniffio                   1.2.0
	sos                       4.1
	spacy                     2.3.2
	SQLAlchemy                1.4.15
	sqlparse                  0.4.1
	srsly                     1.0.2
	ssh-import-id             5.10
	strict-rfc3339            0.7
	swagger-spec-validator    2.7.3
	sympy                     1.5.1
	systemd-python            234
	tabulate                  0.8.9
	tensorboard               2.5.0
	tensorboard-data-server   0.6.1
	tensorboard-plugin-wit    1.8.0
	tensorflow                2.5.0
	tensorflow-estimator      2.5.0
	termcolor                 1.1.0
	terminado                 0.10.1
	testpath                  0.5.0
	text-unidecode            1.3
	thinc                     7.4.1
	threadpoolctl             2.1.0
	tokenizers                0.8.1rc2
	torch                     1.5.1
	tornado                   6.1
	tqdm                      4.46.0
	traitlets                 5.0.5
	transformers              3.1.0
	Twisted                   18.9.0
	typing-extensions         3.7.4.3
	ubuntu-advantage-tools    27.0
	ufw                       0.36
	unattended-upgrades       0.1
	uritemplate               3.0.1
	urllib3                   1.25.8
	urwid                     2.0.1
	wadllib                   1.3.3
	wasabi                    0.8.0
	wcwidth                   0.2.5
	webcolors                 1.11.1
	webencodings              0.5.1
	websocket-client          1.1.0
	Werkzeug                  2.0.1
	wheel                     0.36.2
	widgetsnbextension        3.5.1
	wrapt                     1.12.1
	xkit                      0.0.0
	zipp                      1.0.0
	zope.component            4.3.0
	zope.event                4.4
	zope.hookable             5.0.0
	zope.interface            4.7.1
</pre>
</details>

<details><summary>Command Line Output</summary>
<pre>
➜  ~ [W 2021-06-18 18:50:24.121 LabApp] 'allow_origin' has moved from NotebookApp to ServerApp. This config will be passed to ServerApp. Be sure to update your config before our next release.
[I 2021-06-18 18:50:24.128 ServerApp] jupyterlab | extension was successfully linked.
[I 2021-06-18 18:50:24.153 LabApp] JupyterLab extension loaded from /home/t/.local/lib/python3.8/site-packages/jupyterlab
[I 2021-06-18 18:50:24.153 LabApp] JupyterLab application directory is /home/t/.local/share/jupyter/lab
[I 2021-06-18 18:50:24.156 ServerApp] jupyterlab | extension was successfully loaded.
[I 2021-06-18 18:50:24.157 ServerApp] Serving notebooks from local directory: /home/t/dev/jupyter
[I 2021-06-18 18:50:24.157 ServerApp] Jupyter Server 1.8.0 is running at:
[I 2021-06-18 18:50:24.157 ServerApp] https://00.00.00.00:5000/lab
[I 2021-06-18 18:50:24.157 ServerApp]     https://127.0.0.1:5000/lab
[I 2021-06-18 18:50:24.157 ServerApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
[I 2021-06-18 18:50:29.245 ServerApp] 302 GET / (95.208.11.26) 1.22ms
[I 2021-06-18 18:50:32.447 LabApp] Build is up to date
[W 2021-06-18 18:50:32.986 ServerApp] 404 GET /terminals/websocket/3 (95.208.11.26) 23.26ms referer=None
[W 2021-06-18 18:50:33.007 ServerApp] 404 GET /terminals/websocket/4 (95.208.11.26) 1.11ms referer=None
/[I 2021-06-18 18:50:33.997 ServerApp] Kernel started: e1cacb74-92af-4cd9-b8a7-5df27900e58b
neofetch
</pre>
</details>

<details><summary>Browser Output</summary>
<pre>

</pre>
</details>
